src/compiler/csharp_generator.cc

```diff
@@ -289,114 +280,126 @@ std::string GetMethodReturnTypeServer(const MethodDescriptor *method) {
-        "global::Google.Protobuf.MessageExtensions.ToByteArray(arg), "
-        "$type$.Parser.ParseFrom);\n",
-        "fieldname", GetMarshallerFieldName(message), "type",
-        GetClassName(message));
+      vars,
+      "static readonly grpc::Marshaller<$type$> $fieldname$ = "
+      "grpc::Marshallers.Create((arg) => "
+      "global::Google.Protobuf.MessageExtensions.ToByteArray(arg), " //ERR ref to pb
+      "$type$.Parser.ParseFrom);\n"); 
   }
   out->Print("\n");
 }
 
-void GenerateStaticMethodField(Printer *out, const MethodDescriptor *method) {
+void GenerateStaticMethodField(Printer *out, const grpc_generator::Method *method) {
+
+  std::map<grpc::string, grpc::string> vars;
+  vars["fieldname"] = GetMethodFieldName(method);
+  vars["request"] = method->input_type_name();
+  vars["response"] = method->output_type_name();
   out->Print(
+    vars,
       "static readonly grpc::Method<$request$, $response$> $fieldname$ = new "
-      "grpc::Method<$request$, $response$>(\n",
-      "fieldname", GetMethodFieldName(method), "request",
-      GetClassName(method->input_type()), "response",
-      GetClassName(method->output_type()));
+      "grpc::Method<$request$, $response$>(\n");
   out->Indent();
   out->Indent();
-  out->Print("$methodtype$,\n", "methodtype",
-             GetCSharpMethodType(GetMethodType(method)));
-  out->Print("$servicenamefield$,\n", "servicenamefield",
-             GetServiceNameFieldName());
-  out->Print("\"$methodname$\",\n", "methodname", method->name());
-  out->Print("$requestmarshaller$,\n", "requestmarshaller",
-             GetMarshallerFieldName(method->input_type()));
-  out->Print("$responsemarshaller$);\n", "responsemarshaller",
-             GetMarshallerFieldName(method->output_type()));
+  vars.clear();
+  vars["methodtype"] = GetCSharpMethodType(GetMethodType(method));
+  vars["servicenamefield"] = GetServiceNameFieldName();
+  vars["methodname"] = method->name();
+  vars["requestmarshaller"] = GetMarshallerFieldName(method->input_type_name());
+  vars["output_type"] = GetMarshallerFieldName(method->output_type_name());
+
+  out->Print(vars, "$methodtype$,\n");
+  out->Print(vars, "$servicenamefield$,\n");
+  out->Print(vars, "\"$methodname$\",\n");
+  out->Print(vars, "$requestmarshaller$,\n");
+  out->Print(vars, "$responsemarshaller$);\n");
   out->Print("\n");
   out->Outdent();
   out->Outdent();
 }
 
+
 void GenerateServiceDescriptorProperty(Printer *out,
-                                       const ServiceDescriptor *service) {
+                                       const grpc_generator::Service *service) {
   std::ostringstream index;
-  index << service->index();
+  index << -1; //ERR service->index();
   out->Print("/// <summary>Service descriptor</summary>\n");
   out->Print(
-      "public static global::Google.Protobuf.Reflection.ServiceDescriptor "
+      "public static global::Google.Protobuf.Reflection.ServiceDescriptor "//ERR ref to pb
```

Comment:
Same here. Have a look at the C++ PR I linked in the main discussion on this PR, which creates an interface that supplies Method/Service descriptors/names etc generically.